how-to/register-with-dock/client/src/dock.ts

```diff
@@ -0,0 +1,352 @@
+			if (payload.callerType === CustomActionCallerType.CustomButton && currentView) {
+				await addToFavorites();
+			}
+		},
+		"favorite-remove": async (payload): Promise<void> => {
+			// The favorite remove is triggered from the button at the top of the browser window
+			if (payload.callerType === CustomActionCallerType.CustomButton && currentView) {
+				await removeFromFavorites();
+			}
+		}
+	};
+}
+
+/**
+ * Create the configuration for the dock, including a drop down for favorites if there are any.
+ * @returns The dock configuration.
+ */
+function buildDockConfiguration(): DockProviderConfig {
+	const buttons: DockButton[] = [
+		{
+			tooltip: "Add View",
+			iconUrl: "http://localhost:8080/assets/view-add.svg",
+			action: {
+				id: "view-add"
+			}
+		},
+		{
+			type: DockButtonNames.DropdownButton,
+			tooltip: "Favorites",
+			id: "favorites",
+			iconUrl: "http://localhost:8080/assets/favorites.svg",
+			options:
+				favorites.length > 0
+					? favorites.map((fav) => ({
+							tooltip: fav.title,
+							action: {
+								id: "favorite-open",
+								customData: fav
+							}
+					  }))
+					: [
+							{
+								tooltip: "No favorites",
+								disabled: true,
+								action: {
+									id: "favorite-open"
+								}
+							}
+					  ]
+		}
+	];
+
+	return {
+		title: platformTitle ?? "",
+		icon: platformIcon ?? "",
+		workspaceComponents: ["home", "notifications", "store", "switchWorkspace"],
+		disableUserRearrangement: false,
+		buttons
+	};
+}
+
+/**
+ * Update the dock when something has changed.
+ */
+async function updateDock(): Promise<void> {
+	if (registration) {
+		await registration.updateDockProviderConfig(buildDockConfiguration());
+	}
+}
+
+/**
+ * Update the buttons shown on the browser window.
+ */
+async function updateBrowserButtons(): Promise<void> {
+	const buttons = await getDockBrowserButtons();
+
+	if (buttons) {
+		const platform = getCurrentSync();
+		const browserWindows = await platform.Browser.getAllWindows();
+
+		if (browserWindows.length > 0) {
+			const browserWindow = fin.Window.wrapSync(browserWindows[0].identity);
+			await browserWindow.updateOptions({
+				workspacePlatform: {
+					toolbarOptions: {
+						buttons
+					}
+				}
+			});
+		}
+	}
+}
+
+/**
+ * Open a url in a view.
+ * @param url The view to open.
+ */
+async function openUrl(url: string): Promise<void> {
+	const platform = getCurrentSync();
+
+	// See if we already have a browser window open.
+	let browserWindows = await platform.Browser.getAllWindows();
+
+	let browserWindowTarget: OpenFin.Identity | undefined;
+	if (browserWindows.length > 0) {
+		browserWindowTarget = browserWindows[0].identity;
+	}
+
+	// Open a view
+	const view = await platform.createView({ url }, browserWindowTarget);
+	currentUrl = url;
+	currentView = view;
+	await updateBrowserButtons();
+
+	// If there was no initial window then get it and hook up event listeners
+	if (!browserWindowTarget) {
```

Comment:
So this always targets the first window. Is it worth adding a note that it won't work on the second/third window (the event listener won't be added)? 

Great demo

Reply:
Yes it always reuses the first window for this demo